File: C#/myLeetCode/Binary_Tree/N-ary Tree/559. Maximum Depth of N-ary Tree.cs
// 559. Maximum Depth of N-ary Tree
// https://leetcode.com/problems/maximum-depth-of-n-ary-tree/

using System;
using System.Collections.Generic;

public class Node {
    public int val;
    public IList<Node> children;

    public Node() {}

    public Node(int _val) {
        val = _val;
        children = new List<Node>();
    }

    public Node(int _val, IList<Node> _children) {
        val = _val;
        children = _children;
    }
}

public class Solution {
    public int MaxDepth(Node root) {
        if (root == null) {
            return 0;
        }

        int maxChildDepth = 0;
        foreach (var child in root.children) {
            int childDepth = MaxDepth(child);
            maxChildDepth = Math.Max(maxChildDepth, childDepth);
        }

        return maxChildDepth + 1;
    }
}

public class Program {
    public static void Main() {
        Solution solution = new Solution();

        Node tree = new Node(1, new List<Node> {
            new Node(3, new List<Node> {
                new Node(5),
                new Node(6)
            }),
            new Node(2),
            new Node(4)
        });

        int maxDepth = solution.MaxDepth(tree);
        Console.WriteLine($"Maximum Depth of N-ary Tree: {maxDepth}");
    }
}

File: C++/myLeetCode/Binary_Tree/N-ary Tree/559. Maximum Depth of N-ary Tree.cpp
// 559. Maximum Depth of N-ary Tree
// https://leetcode.com/problems/maximum-depth-of-n-ary-tree/
// clang++ 559.\ Maximum\ Depth\ of\ N-ary\ Tree.cpp -std=c++17
#include "../../printUtils.hpp"

template<typename T>
int maxDepth(NTreeNode<T>* root) {
    if (root == nullptr) return 0;

    int depth = 0;
    for (NTreeNode<T>* child : root->children) {
        depth = std::max(depth, maxDepth(child));
    }

    return depth + 1;
}

int main() {
    // NTreeNode<int>* root = buildNTree<int>({1, -1, 3, 2, 4, -1, 5, 6}, -1);
    NTreeNode<int>* root = buildNTree<int>({1,-1,2,3,4,5,-1,-1,6,7,-1,8,-1,9,10,-1,-1,11,-1,12,-1,13,-1,-1,14}, -1);
    std::cout << "Maximum depth of the N-ary tree: " << maxDepth(root) << std::endl;
    deleteNTree(root);
    
    return 0;
}
